File: ios/Wavepoint/Auth/SpotifyCredentialProvider.swift
import Foundation
import Supabase

struct RefreshedSpotifyToken: Equatable, Sendable, Decodable {
  let accessToken: String
  let refreshToken: String?
  let expiresIn: Int

  enum CodingKeys: String, CodingKey {
    case accessToken = "access_token"
    case refreshToken = "refresh_token"
    case expiresIn = "expires_in"
  }
}

protocol SpotifyTokenRefreshServing: Sendable {
  func refresh(providerRefreshToken: String) async throws -> RefreshedSpotifyToken
}

actor SpotifyCredentialProvider {
  private let tokenStore: any SpotifyTokenStoring
  private let refreshService: any SpotifyTokenRefreshServing

  init(
    tokenStore: any SpotifyTokenStoring,
    refreshService: any SpotifyTokenRefreshServing
  ) {
    self.tokenStore = tokenStore
    self.refreshService = refreshService
  }

  func accessToken(forceRefresh: Bool) async throws -> String {
    guard let storedTokens = try tokenStore.load() else {
      throw SpotifyWebAPIError.authorizationExpired
    }
    guard forceRefresh else { return storedTokens.accessToken }
    guard let refreshToken = storedTokens.refreshToken else {
      throw SpotifyWebAPIError.authorizationExpired
    }

    let refreshed = try await refreshService.refresh(
      providerRefreshToken: refreshToken
    )
    try tokenStore.save(
      SpotifyProviderTokens(
        accessToken: refreshed.accessToken,
        refreshToken: refreshed.refreshToken ?? refreshToken
      )
    )
    return refreshed.accessToken
  }
}

struct SupabaseSpotifyTokenRefreshService: SpotifyTokenRefreshServing {
  private let client: SupabaseClient
  private let functionURL: URL
  private let publishableKey: String
  private let session: URLSession

  init(
    client: SupabaseClient,
    configuration: AppConfiguration,
    session: URLSession = .shared
  ) {
    self.client = client
    functionURL = configuration.supabaseURL
      .appending(path: "functions/v1/spotify-token-refresh")
    publishableKey = configuration.supabasePublishableKey
    self.session = session
  }

  func refresh(providerRefreshToken: String) async throws -> RefreshedSpotifyToken {
    let supabaseSession = try await client.auth.session
    var request = URLRequest(url: functionURL)
    request.httpMethod = "POST"
    request.setValue(
      "Bearer \(supabaseSession.accessToken)",
      forHTTPHeaderField: "Authorization"
    )
    request.setValue(publishableKey, forHTTPHeaderField: "apikey")
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    request.httpBody = try JSONEncoder().encode(
      RefreshRequest(refreshToken: providerRefreshToken)
    )

    let (data, response) = try await session.data(for: request)
    guard let response = response as? HTTPURLResponse else {
      throw SpotifyWebAPIError.invalidResponse
    }
    guard response.statusCode == 200 else {
      if response.statusCode == 401 {
        throw SpotifyWebAPIError.authorizationExpired
      }
      throw SpotifyWebAPIError.httpStatus(response.statusCode)
    }

    do {
      return try JSONDecoder().decode(RefreshedSpotifyToken.self, from: data)
    } catch {
      throw SpotifyWebAPIError.invalidData
    }
  }
}

private struct RefreshRequest: Encodable {
  let refreshToken: String

  enum CodingKeys: String, CodingKey {
    case refreshToken = "refresh_token"
  }
}

File: android/app/src/main/java/ai/mapier/swipe/auth/SpotifyCredentialProvider.kt
package ai.mapier.swipe.auth

import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.functions.functions
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.HttpHeaders
import io.ktor.http.headersOf
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

data class RefreshedSpotifyToken(
  val accessToken: String,
  val refreshToken: String?,
  val expiresIn: Int,
)

interface SpotifyTokenRefreshService {
  suspend fun refresh(providerRefreshToken: String): RefreshedSpotifyToken
}

class SpotifyAuthorizationExpiredException : Exception("Spotify authorization expired")

class SpotifyCredentialProvider(
  private val tokenStore: SpotifyTokenStoring,
  private val refreshService: SpotifyTokenRefreshService,
) {
  suspend fun accessToken(forceRefresh: Boolean): String {
    val stored = tokenStore.load() ?: throw SpotifyAuthorizationExpiredException()
    if (!forceRefresh) return stored.accessToken
    val refreshToken = stored.refreshToken ?: throw SpotifyAuthorizationExpiredException()
    val refreshed = refreshService.refresh(refreshToken)
    tokenStore.save(
      SpotifyProviderTokens(
        accessToken = refreshed.accessToken,
        refreshToken = refreshed.refreshToken ?: refreshToken,
      ),
    )
    return refreshed.accessToken
  }
}

class SupabaseSpotifyTokenRefreshService(
  private val client: SupabaseClient,
  private val json: Json = Json { ignoreUnknownKeys = true },
) : SpotifyTokenRefreshService {
  override suspend fun refresh(providerRefreshToken: String): RefreshedSpotifyToken {
    val response = client.functions.invoke(
      function = "spotify-token-refresh",
      body = RefreshRequest(providerRefreshToken),
      headers = headersOf(HttpHeaders.ContentType, ContentType.Application.Json.toString()),
    )
    val decoded = json.decodeFromString<RefreshResponse>(response.bodyAsText())
    return RefreshedSpotifyToken(
      accessToken = decoded.accessToken,
      refreshToken = decoded.refreshToken,
      expiresIn = decoded.expiresIn,
    )
  }
}

@Serializable
private data class RefreshRequest(
  @SerialName("refresh_token") val refreshToken: String,
)

@Serializable
private data class RefreshResponse(
  @SerialName("access_token") val accessToken: String,
  @SerialName("refresh_token") val refreshToken: String? = null,
  @SerialName("expires_in") val expiresIn: Int,
)
